### ISSUE TITLE
Default ignore minified files

### ISSUE THREAD

**Issue by jayvdb** [coala/coala#3131]
minified files, such as `**.min.js`, look terrible as the UI prints the entire line (https://github.com/coala/coala/issues/3130), and these files typically violate every rule in the book anyway.

**Comment by nemaniarjun** [comment]
I would like to give this a shot

**Comment by jayvdb** [comment]
@nemaniarjun , you're still working on two other tasks.  This task could easily have more analysis done (e.g. what other languages have generated files with standard suffix?), and design (where do we put this ignore list, which will grow?), which you could think about and post here to this issue.

**Comment by ashwini0529** [comment]
Although I don't have much idea about it. But I can try to work on it.

What I think could work is : 

We can create a file with the list of all the extensions that are to be excluded like `.min.css`, `.min.js` etc.
Whenever any bear runs, it would first check the file extension with the list, and if it is not in that file, it will start scanning it, otherwise it will skip the file.

**Comment by Makman2** [comment]
don't we have default ignores? We could place this inside our global default coafile (if we still have it, don't remember...)

**Comment by nemaniarjun** [comment]
We can keep the list of globs in `ConfigurationGathering.py`
We can modify `icollect(file_paths, ignored_globs=None)`

**Comment by Makman2** [comment]
@nemaniarjun so you wanna change the default parameter `ignored_globs`? This is bad, `icollect` should work as modular as possible, this is analysis specific for java bears, though it's still a useful default (that's why we should place it somewhere else in coala)^^

**Comment by nemaniarjun** [comment]
No I meant we can make modifications to icollect
To ignore a list of globs we have stored in `configurationsetting` , `icollect` works on all bears to `unselect` user defined `ignore` globs right?  Without changing its current functionality.

**Comment by jayvdb** [comment]
We shouldnt be changing the core.
If the user has selected their own `ignore`, it should be used, *and* no voodoo default ignores used.
However, coala should have a `CONSTANT` fallback of sane default ignores.

So I believe we need a constant stored somewhere, and one of the callers of `icollect` needs to set the "user" ignore list to be the default ignore list, only when the user did not provide any.

**Comment by nemaniarjun** [comment]
```

@yield_once
def icollect(file_paths, ignored_globs=None):
    """
    Evaluate globs in file paths and return all matching files.

    :param file_paths:    File path or list of such that can include globs
    :param ignored_globs: List of globs to ignore when matching files
    :return:              Iterator that yields tuple of path of a matching
                          file, the glob where it was found
    """
    if isinstance(file_paths, str):
        file_paths = [file_paths]

    for file_path in file_paths:
        for match in iglob(file_path):
            if not ignored_globs or not fnmatch(match, ignored_globs):
                yield match, file_path
```

**Comment by jayvdb** [comment]
@nemaniarjun, try it.

**Comment by Makman2** [comment]
@sils sorry that I need to ask you, but do we still have the system wide coafile with a system wide default-section?

@jayvdb @nemaniarjun I think this default ignore would belong there, then we need to change no code at all (so we don't need any fallbacks)

**Comment by jayvdb** [comment]
@Makman2, absolutely agree.  There is a `coalib/default_coafile` .. ?

**Comment by sils** [comment]
yes :)

**Comment by nemaniarjun** [comment]
About the default coafile, does it work in addition to the coafile? Or does it work only when the user has not made a coafile? 
If it's the first then we need to provide a way to disable the default coafile. In order to run coala on  .min files.

**Comment by jayvdb** [comment]
@nemaniarjun , you can determine that via the code and tests.  I've created an issue for documenting it at https://github.com/coala/coala/issues/3388 .

**Comment by Makman2** [comment]
@nemaniarjun they should work both together I believe, especially the default sections should be merged of the system coafile and use coafile.

**Comment by nemaniarjun** [comment]
Should we change the name of `default_coafile` in coalib to `system_coafile`?
It will avoid confusion due to the following lines in `Constants.py`
```
system_coafile = os.path.join(coalib_root, 'default_coafile')

user_coafile = os.path.join(os.path.expanduser('~'), '.coarc')

default_coafile = '.coafile'
```

**Comment by Makman2** [comment]
open an issue for changing the naming convention of this file^^

**Comment by nemaniarjun** [comment]
Uhmm need a bit of help

```
arjun@Arjun:~/coala$ l
a.min.js  coala/  coala-bears/
arjun@Arjun:~/coala$ coala --files *.js --bears SpaceConsistencyBear
[WARNING][02:38:16] The default coafile '.coafile' was not found. You can generate a configuration file with your current options by adding the `--save` flag or suppress any use of config files with `-I`.
Please enter a value for the setting "use_spaces" (True if spaces are to be used instead of tabs.) needed by SpaceConsistencyBear for section "Default": 
True
Executing section Default...

a.min.js
|   2| »   a
|    | [NORMAL] SpaceConsistencyBear:
|    | Line contains following spacing inconsistencies:
|    | - No newline at EOF.
|    | - Tabs used instead of spaces.
|----|    | /home/arjun/coala/a.min.js
|    |++++| /home/arjun/coala/a.min.js
|   1|   1|     a
|   2|    |-	a
|    |   2|+    a
|    | *0: Do nothing
|    |  1: Open file(s)
|    |  2: Apply patch
|    |  3: Print more info
|    |  4: Add ignore comment
|    | Enter number (Ctrl-D to exit): 0
arjun@Arjun:~/coala$ cat coala/coalib/default_coafile 
[default]
ignore = *.min.js
arjun@Arjun:~/coala$ 
```

**Comment by Makman2** [comment]
You execute `coala` from your system, not from the repo root :) You need to modify the installed `default_coafile` (wherever it lies), or maybe it's also possible to execute `./coala` from repo root, maybe it grabs the default-coafile from the repo^^

**Comment by nemaniarjun** [comment]
I edited and installed the coala repo and no I dont think it takes it from the repo. 
It uses the installed one:
```
system_coafile = os.path.join(coalib_root, 'default_coafile')
```

**Comment by nemaniarjun** [comment]
Also doesnt work when I run it in the coala repo

**Comment by Makman2** [comment]
then you need to modify the installed one, you could `python3 setup.py install`^^

**Comment by nemaniarjun** [comment]
```
arjun@Arjun:~/coala/coala$ cat /usr/local/lib/python3.5/dist-packages/coala-0.10.0.dev99999999999999-py3.5.egg/coalib/default_coafile 
[default]
ignore = *.min.js
arjun@Arjun:~/coala/coala$ coala
Executing section Default...
Executing section python...
Executing section autopep8...
Executing section linelength...
Executing section DOCS...
Executing section commit...
Executing section rst...
Executing section text...

a.min.js
|   2| »   a
|    | [NORMAL] SpaceConsistencyBear:
|    | Line contains following spacing inconsistencies:
|    | - No newline at EOF.
|    | - Tabs used instead of spaces.
|----|    | /home/arjun/coala/coala/a.min.js
|    |++++| /home/arjun/coala/coala/a.min.js
|   1|   1|     a
|   2|    |-	a
|    |   2|+    a
|    | *0: Do nothing
|    |  1: Open file(s)
|    |  2: Apply patch
|    |  3: Print more info
|    |  4: Add ignore comment
|    | Enter number (Ctrl-D to exit): 0
arjun@Arjun:~/coala/coala$ 

```

**Comment by nemaniarjun** [comment]
btw, this is how my .coafile looks like: 
```
[text]
bears = SpaceConsistencyBear
files =
    LICENSE, ISSUE_TEMPLATE, PULL_REQUEST_TEMPLATE.md, MANIFEST.in, *.js,
    **.sh, .travis.yml, circle.yml, setup.cfg, ./docs/_static/custom.css
ignore = node_modules/**
```

**Comment by nemaniarjun** [comment]
I think the `load_configuration` in `ConfigurationGathering.py` has some flaw..
Even in the tests, the `user_coafile` is not tested together with `default_coafile`

**Comment by Makman2** [comment]
yeah yeah wait if you override `ignore` you lose the ignore from the base coafile, you need to remove your `ignore`, we don't merge list values from the base file^^ You could file another issue to discuss the behaviour for these cases^^

**Comment by Makman2** [comment]
I'm not sure if we already support that, but you could try `ignore += your-stuff`^^

**Comment by nemaniarjun** [comment]
It doesn't work...

**Comment by sils** [comment]
we don't support that yet

**Comment by jayvdb** [comment]
This bug is about a sensible default ignore list.
See https://github.com/coala/coala/issues/3131#issuecomment-269079230

If the user customises the `ignore`, they can add minified paths to their custom `ignore`.

Smartly merging user ignore and system ignore is a separate problem.

Can we get a status update about this issue.  It looks like it is a really simple addition which could make it into 0.10.

**Comment by adhikasp** [comment]
This would be easy addition *if* https://github.com/coala/coala/pull/3337/files merged (the ability to append value in coala configuration).

**Comment by sils** [comment]
@adhikasp that functionality was already merged, that PR is just a simplification of the confparser which it is using.

**Comment by userzimmermann** [comment]
@sils Anyway... @adhikasp Thanks for promoting #3337 :)

**Comment by sils** [comment]
@userzimmermann not true? I didn't mean to say it's not useful, just saying that this issue isn't blocked which it isn't right?

**Comment by userzimmermann** [comment]
@sils Righty right! And true :)

**Comment by jayvdb** [comment]
If https://github.com/ggreer/the_silver_searcher/pull/232 is added to `ag`, we could have an optional `ag` mode which gets the initial list of files from `ag` (and defaults back to normal mode if a user doesnt have it installed).

But, [`ack`](https://beyondgrep.com/install/) and [`pt`](https://github.com/monochromegane/the_platinum_searcher) and [`sift`](https://github.com/svent/sift) already have this functionality, and could be used now.
